ethcore/src/miner/stratum.rs

```diff
@@ -0,0 +1,198 @@
+
+/// Job dispatcher for stratum service
+pub struct StratumJobDispatcher {
+	last_work: RwLock<Option<(H256, U256, u64)>>,
+	seed_compute: Mutex<SeedHashCompute>,
+	client: Weak<Client>,
+	miner: Weak<Miner>,
+}
+
+impl JobDispatcher for StratumJobDispatcher {
+	fn initial(&self) -> Option<String> {
+		// initial payload may contain additional data, not in this case
+		self.job()
+	}
+
+	fn job(&self) -> Option<String> {
+		self.with_core(|client, miner| {
+			if let Some((pow_hash, difficulty, number)) = miner.map_sealing_work(&*client, |b| {
+				let pow_hash = b.hash();
+				let number = b.block().header().number();
+				let difficulty = b.block().header().difficulty();
+
+				(pow_hash, *difficulty, number)
+			}) {
+				*self.last_work.write() = Some((pow_hash, difficulty, number));
+				Some(self.payload(pow_hash, difficulty, number))
+			} else { None }
+		})
+	}
+
+	fn submit(&self, payload: Vec<String>) {
+		if payload.len() != 3 {
+			warn!(target: "stratum", "submit_work: invalid work submit request({:?})", payload);
+		}
+		else {
+			trace!(target: "stratum", "submit_work: {} {} {}", &payload[0], &payload[1], &payload[2]);
+		}
+
+		let nonce = match H64::from_str(clean_0x(&payload[0])) {
+			Ok(nonce) => nonce,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work ({}): invalid nonce ({:?})", &payload[0], e);
+				return;
+			}
+		};
+
+		let pow_hash = match H256::from_str(clean_0x(&payload[1])) {
+			Ok(pow_hash) => pow_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work ({}): invalid hash ({:?})", &payload[1], e);
+				return;
+			}
+		};
+
+		let mix_hash = match H256::from_str(clean_0x(&payload[2])) {
+			Ok(mix_hash) => mix_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work ({}): invalid mix-hash ({:?})",  &payload[2], e);
+				return;
+			}
+		};
+
+		trace!(target: "stratum", "submit_work: Decoded: nonce={}, pow_hash={}, mix_hash={}", nonce, pow_hash, mix_hash);
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
+
+		let seal = vec![encode(&mix_hash).to_vec(), encode(&nonce).to_vec()];
+		if let Err(e) = miner.submit_seal(&*client, pow_hash, seal) {
+			warn!(target: "stratum", "submit_work error: {:?}", e);
+		};
+	}
+}
+
+impl StratumJobDispatcher {
+	/// New stratum job dispatcher given the miner and client
+	fn new(miner: Weak<Miner>, client: Weak<Client>) -> StratumJobDispatcher {
+		StratumJobDispatcher {
+			seed_compute: Mutex::new(SeedHashCompute::new()),
+			last_work: RwLock::new(None),
+			client: client,
+			miner: miner,
+		}
+	}
+
+	/// Serializes payload for stratum service
+	fn payload(&self, pow_hash: H256, difficulty: U256, number: u64) -> String {
+		// TODO: move this to engine
+		let target = Ethash::difficulty_to_boundary(&difficulty);
+		let seed_hash = &self.seed_compute.lock().get_seedhash(number);
+		let seed_hash = H256::from_slice(&seed_hash[..]);
+		format!(
```

Comment:
it's basically one request message and one reply message, so i don't feel it is neccessary